ntbs-service/Pages/ServiceDirectory/SearchResults.cshtml

```diff
@@ -11,82 +11,143 @@
-                                                @tbService.PHEC?.Name
-                                            </a>
-                                            <span> > </span>
-                                            <a class="link-without-visited-state"
-                                               href=@($"/ServiceDirectory/Region/{tbService.PHECCode}#accordion-heading-{tbService.Code}")>
-                                                @tbService.Name
-                                            </a>
-                                        </span>
-                                    </nhs-summary-list-row-value>
-                                </nhs-summary-list-row>
-                            }
-                            @foreach (var region in Model.AllPhecs.Where(p => user.AdGroups != null && user.AdGroups.Split(",").Contains(p.AdGroup)))
-                            {
+                @foreach (var region in Model.RegionSearchResults)
+                {
+                    <li>
+                        <div class="directory-search-result-container non-user-search-result" id="">
+                            <b>Region:</b>
+                            <a class="link-without-visited-state" href=@($"/ServiceDirectory/Region/{region.Code}")>
+                                @region.Name
+                            </a>
+                        </div>
+                    </li>
+                }
+            }
+            
+            @if (Model.TbServiceSearchResults is not null)
+            {
+                @foreach (var tbService in Model.TbServiceSearchResults)
+                {
+                    <li>
+                        <div class="directory-search-result-container non-user-search-result" id="">
+                            <b>Service:</b>
+                            <a class="link-without-visited-state" href=@($"/ServiceDirectory/Region/{tbService.PHEC.Code}")>
+                                @tbService.PHEC?.Name
+                            </a>
+                            <span> > </span>
+                            <a class="link-without-visited-state"
+                               href=@($"/ServiceDirectory/Region/{tbService.PHECCode}#accordion-heading-{tbService.Code}")>
+                                @tbService.Name
+                            </a>
+                        </div>
+                    </li>
+                }
+            }
+            
+            @if (Model.HospitalSearchResults is not null)
+            {
+                @foreach (var hospital in Model.HospitalSearchResults)
+                {
+                    <li>
+                        <div class="directory-search-result-container non-user-search-result" id="">
+                            <b>Hospital:</b>
+                            <a class="link-without-visited-state" href=@($"/ServiceDirectory/Region/{hospital.TBService.PHEC.Code}")>
+                                @hospital.TBService?.PHEC?.Name
```

Comment:
When you fetch the hospitals you're not including the TB service or the region, so surely this is always empty?

Reply:
It's not, but now that you mention it I'm not sure why it isn't

Reply:
Having looked it up, I think it's because Lazy Loading has been enabled by marking the TBservice and PHEC properties on Hospital as virtual

Reply:
Hmm, I thought lazy loading was disabled by default in EF Core and you had to explicitly enable it (which I don't think we're doing - because lazy loading is a terrible thing). I'm going to investigate this a bit.

Reply:
Right, it took me a little while, but I've finally figured this out! What's happening here is:
- you're searching in the order: user, region, TB service, hospital
- when searching by region, you're pulling all of the PHECs into the context
- when searching by TB service, you're pulling all of the TB services into the context
- when searching by hospital, you're pulling all of the hospitals into the context
- the context is correlating the IDs on the objects you've loaded from the database with the IDs of the objects it has already fetched

This works, but it's very fragile! For example, if you change the way that the TB service search works, so that it only fetches the services that match the keywords, then suddenly this (completely unrelated bit of code) will stop working!

You should do one of the following:
- add an include statement to your query in the repository
- be much more explicit about what you're doing, e.g. by explicitly fetching all of the PHECs and TB services in your `SearchResults` class and adding a comment to say why you're doing it

Reply:
This has thoroughly surprised me - to test I just debugged these four lines (I've purposefully changed the order)
`var hospitalsToDisplay = await _hospitalSearchService.OrderQueryableAsync(searchKeywords);
 var usersToDisplay = await _userSearchService.OrderQueryableAsync(searchKeywords);
 var regionsToDisplay = await _regionSearchService.OrderQueryableAsync(searchKeywords);
 var tbServicesToDisplay = await _tbServiceSearchService.OrderQueryableAsync(searchKeywords);`
and TBService on hopsitalsToDisplay is empty until line 4 is evaluated so you are definitely right.

Is there somewhere I can read up on why this behaviour is occurring? I'm not sure how to find the infomation I'm looking for

Reply:
I've had a quick look, but I'm afraid I haven't found anything. FWIW, my approach to this was:
- read the documentation [here](https://docs.microsoft.com/en-us/ef/core/querying/related-data/) about eager vs lazy loading in EF core and check that my understanding was correct
- debug the code and observe that the TB service property was present on the object immediately after being fetched
- debug the query being generated by EF, using the `ToQueryString()` method - which told me that the query was not joining to the TB service table, so the data had to be coming from elsewhere
- scratch my head and think for a while
- realise that TB services may already be in memory because of the previous search, and then change the order of the searches to confirm my theory